### PR TITLE
Add mouse control features for touch devices

### PR DESCRIPTION
**Pull request by yaniv15**
This PR add mouse control features for touch devices like android and IOS.
Since android and IOS doesn't support typical mouse events like mousedown, mouseup they where mocked by touchstart, touchend and other touch events.

currently supported actions:
single tap: left click.
double tap: double left click.
long tap: right click.
two finger swipe: scroll.

https://github.com/tiny-pilot/tinypilot/assets/100175905/382a0488-a4f3-4b7e-a86e-58ee5694042e

Tested on android and ios browsers where tiinypilot is controlling ubuntu system.
Note: Disabled the existing mobile keyboard as it is interfering with touch events. A new PR will be raised for mobile-keyboard.
<a data-ca-tag href="https://codeapprove.com/pr/tiny-pilot/tinypilot/1387">Review on CodeApprove</a>

### CODE REVIEW

File: app/templates/custom-elements/remote-screen.html [hunk 1]
```diff
@@ -158,34 +158,11 @@
           this.shadowRoot.addEventListener("touchend", () => {
```
Comment (mtlynch) [inline app/templates/custom-elements/remote-screen.html:158]: When I tested this on Android + Firefox, it treats separate touch events as a mouse button that's being held down:

https://github.com/tiny-pilot/tinypilot/assets/7783288/cab78f08-6f53-4ea4-a6cd-718d8c101cbb

**Review by mtlynch — CHANGES_REQUESTED** [review]
Thanks for contributing this! I've added a few notes.

File: app/static/css/style.css [hunk 2]
```diff
@@ -86,6 +87,7 @@ body {
 
 html {
   font-family: "Overpass", sans-serif;
+  -webkit-touch-callout: none;
```
[the remark this answers is not in the document]
Comment (yaniv15) [reply app/static/css/style.css:90]: This property disables the default callout which gets triggered on long press

File: app/templates/custom-elements/remote-screen.html (hunk 1 above)
Comment (yaniv15) [reply app/templates/custom-elements/remote-screen.html:158]: Looks like mouseup event is not being fired on firefox, I'll debug this tonight.

File: app/templates/custom-elements/remote-screen.html [hunk 3]
```diff
@@ -207,8 +184,94 @@
...
+              //considered as a long touch
+              screenElement.dispatchEvent(mouseEvent);
+            }
+          });
+
+          let lastPos = null;
+          screenElement.addEventListener("touchmove", (evt) => {
+            isPointerMoving = true;
+            const numberOfPoints = evt.touches.length;
+            if (numberOfPoints == 2) {
```
[the remark this answers is not in the document]
Comment (yaniv15) [reply app/templates/custom-elements/remote-screen.html:262]: Sure.

File: app/templates/custom-elements/remote-screen.html [hunk 4]
```diff
@@ -207,8 +184,94 @@
             this.rateLimitedMouse.onMouseMove(evt);
           });
           screenElement.addEventListener("mousedown", (evt) => {
-            this.rateLimitedMouse.onMouseDown(evt);
+            //mousedown is not getting any buttons attibutes on ios - a naive fix.
+            //assuming mousedown always have buttons attribute set and default is 1
```
[the remark this answers is not in the document]
Comment (yaniv15) [reply app/templates/custom-elements/remote-screen.html:188]: Yes, a normal touch is automatically detected as mousedown and mouseup events on both ios and android.

File: app/templates/custom-elements/remote-screen.html [hunk 5]
```diff
@@ -207,8 +184,94 @@
             this.rateLimitedMouse.onMouseMove(evt);
           });
           screenElement.addEventListener("mousedown", (evt) => {
-            this.rateLimitedMouse.onMouseDown(evt);
+            //mousedown is not getting any buttons attibutes on ios - a naive fix.
```
[the remark this answers is not in the document]
Comment (yaniv15) [reply app/templates/custom-elements/remote-screen.html:187]: I'll include this.

File: app/templates/custom-elements/remote-screen.html [hunk 6]
```diff
@@ -158,34 +158,11 @@
           this.shadowRoot.addEventListener("touchend", () => {
             isTouchScreen = true;
           });
-          this.shadowRoot.addEventListener("click", () => {
```
[the remark this answers is not in the document]
Comment (yaniv15) [reply app/templates/custom-elements/remote-screen.html:161]: This click is trying to focus mobile-keyboard-input which will pull-up a keyboard on a simple click. As it is interfering with the touch events I got to remove it. I'll raise a new PR for keyboard mocks once this is approved.

File: app/templates/custom-elements/remote-screen.html (hunk 1 above)
Comment (yaniv15) [reply app/templates/custom-elements/remote-screen.html:163]: Fixed

File: app/templates/custom-elements/remote-screen.html (hunk 3 above)
[the remark this answers is not in the document]
Comment (yaniv15) [reply app/templates/custom-elements/remote-screen.html:279]: Addrd.

**Comment by mtlynch** [comment]
@yaniv15 - Thanks for the updates! We've integrated ESLint into the build now (#1390), which will flag some errors in this PR automatically. Can you merge the latest changes from `master` into this PR and fix the issues ESLint flags?

File: app/templates/custom-elements/remote-screen.html (hunk 1 above)
Comment (mtlynch) [reply app/templates/custom-elements/remote-screen.html]: Cool, good progress! I still can't click and drag on Android. Is that expected?

https://github.com/tiny-pilot/tinypilot/assets/7783288/e0e90f0b-0152-415a-871f-303eb08461fb

Comment (yaniv15) [reply app/templates/custom-elements/remote-screen.html]: I forgot that functionality, I'll try to add it in a new PR.

File: app/static/css/style.css [hunk 7]
```diff
@@ -68,6 +68,7 @@ body {
   text-align: center;
   margin: 0 auto;
   padding-top: 55px; /* Account for menu bar plus some extra spacing */
+  padding-bottom: 40px; /* Account for status bar plus some extra spacing */
```
Comment (mtlynch) [inline app/static/css/style.css:71]: Is this part of these changes? It doesn't seem related.

File: app/templates/custom-elements/remote-screen.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (mtlynch) [reply app/templates/custom-elements/remote-screen.html:161]: Deleting existing code will break existing functionality, so we need to keep it.

File: app/templates/custom-elements/remote-screen.html [hunk 8]
```diff
@@ -153,29 +158,15 @@
...
-          });
+          //needed for later keyboard PR
+          // Detect whether this is a touchscreen device .
+          // let isTouchScreen = false;
+          // this.shadowRoot.addEventListener("touchend", () => {
+          //   isTouchScreen = true;
+          // });
+          // this.shadowRoot.addEventListener("click", () => {
+          //   if (isTouchScreen) {
+          //     this.shadowRoot.getElementById("mobile-keyboard-input").focus();
```
Comment (mtlynch) [inline app/templates/custom-elements/remote-screen.html:169]: Can we delete this and pull it out of source history when we need it? We don't keep commented out code in the repo.

File: app/templates/custom-elements/remote-screen.html [hunk 9]
```diff
@@ -199,6 +190,107 @@
           screenElement.addEventListener("mousedown", (evt) => {
             this.rateLimitedMouse.onMouseDown(evt);
           });
+          //-------------------------------------------------------------------------------
+          var isPointerMoving = false;
```
Comment (mtlynch) [inline app/templates/custom-elements/remote-screen.html:194]: We're adding a lot of touch-specific logic to the remote screen component, which I'd like to avoid because there's already a lot of complexity here.

Could we refactor this into a separate file like `touchHandler.js` (similar to `app/static/js/mouse.js`). And then, we could reduce the logic in `remote-screeen.html` to something like:

```javascript
screenElement.addEventListener("touchstart", (evt) => {
  this.touchHandler.onTouchStart(evt);
});
screenElement.addEventListener("touchend", (evt) => {
  this.touchHandler.onTouchEndevt);
});
...
```

That way, we could encapsulate all the variables that we only need for touch logic to just live in `touchHandler.js`.

File: app/templates/custom-elements/remote-screen.html [hunk 10]
```diff
@@ -199,6 +190,107 @@
...
+          var touchStartTimestamp = null;
+          var longTouchDuration = 600; // Time in milliseconds for a touch to be considered a long touch
```
Comment (mtlynch) [inline app/templates/custom-elements/remote-screen.html:196]: Is this an arbitrary decision we're making? Or does this match a convention elsewhere? We should just document whether it's us making the decision or if this is defined somewhere.

File: app/templates/custom-elements/remote-screen.html [hunk 11]
```diff
@@ -199,6 +190,107 @@
...
+                bubbles: true,
+                cancelable: true,
+                view: window,
+                button: 0,
+                buttons: 1,
+                clientX: touchClientX,
+                clientY: touchClientY,
+              });
+              screenElement.dispatchEvent(mouseEvent);
+              //and an immediate mouse up event.
```
Comment (mtlynch) [inline app/templates/custom-elements/remote-screen.html:246]: Can we keep the comments full sentences and sentence-style capitalization? Also, our convention is to include a single space after the `//`.

(ditto throughout)

**Review by mtlynch — CHANGES_REQUESTED** [review]
Thanks! I think it needs a bit more refactoring to match our standards, but this is good progress.

File: app/templates/custom-elements/remote-screen.html (hunk 9 above)
Comment (yaniv15) [reply app/templates/custom-elements/remote-screen.html]: refactored.

File: app/static/css/style.css (hunk 7 above)
Comment (yaniv15) [reply app/static/css/style.css]: Removed.

File: app/templates/custom-elements/remote-screen.html (hunk 8 above)
Comment (yaniv15) [reply app/templates/custom-elements/remote-screen.html]: Removed.

File: app/templates/custom-elements/remote-screen.html (hunk 10 above)
Comment (yaniv15) [reply app/templates/custom-elements/remote-screen.html]: That is an arbitrary decision.

**Comment by mtlynch** [comment]
@yaniv15 - There are still outstanding comments from my last review about deleting code we need to keep and writing comments with full punctuation.

**Comment by yaniv15** [comment]
> @yaniv15 - There are still outstanding comments from my last review about deleting code we need to keep and writing comments with full punctuation.

I guess the only functionality that piece of code provide is to pull up a keyboard when the user touches the screen. If we don't remove this code, the keyboard will be always shown which covers the screen and user will nut be able to interact using touch events.

**Comment by AltF5** [comment]
@yaniv15 … I’m just a passer-by but I wanted to recommend implementation or similar to that of Jump Desktop, if you have ever used on iOS where you move the cursor around with a circle ⭕️ and basically click in the circle to click, or elsewhere to pop a keyboard. It’s better then any other remote mouse I’ve tried like RVNC or TV. 

Just wanted to drop the suggestion for future improvement

**Comment by mtlynch** [comment]
>>There are still outstanding comments from my last review about deleting code we need to keep and writing comments with full punctuation.

>I guess the only functionality that piece of code provide is to pull up a keyboard when the user touches the screen. If we don't remove this code, the keyboard will be always shown which covers the screen and user will nut be able to interact using touch events.

Doesn't that break the previous functionality, though? How does a mobile user pull up a keyboard and type under this implementation?

**Comment by mtlynch** [comment]
@yaniv15 - I'm closing this due to inactivity, but feel free to reopen when you're ready to resume.
